pkg/scheduler/plugins/reservation/plugin.go

```diff
@@ -0,0 +1,530 @@
+	}
+
+	// update reservation allocation of the pod
+	newPod := pod.DeepCopy()
+	apiext.SetReservationAllocated(newPod, target)
+	patchBytes, err := generatePodPatch(pod, newPod)
+	if err != nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+	err = retry.OnError(
+		retry.DefaultRetry,
+		errors.IsTooManyRequests,
+		func() error {
+			_, err1 := p.handle.ClientSet().CoreV1().Pods(pod.Namespace).
+				Patch(ctx, pod.Name, apimachinerytypes.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+			if err1 != nil {
+				klog.Error("failed to patch Pod %s/%s, patch %v, err: %v",
+					pod.Namespace, pod.Name, string(patchBytes), err1)
+			}
+			return err1
+		})
+	if err != nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+
+	return nil
+}
+
+// Bind fake binds reserve pod and mark corresponding reservation as Available.
+// NOTE: This Bind plugin should get called before DefaultBinder; plugin order should be configured.
+func (p *Plugin) Bind(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeName string) *framework.Status {
+	// 1. if the pod is a reserve pod
+	if IsReservePod(pod) {
+		klog.V(3).InfoS("Attempting to fake bind reserve pod to node", "pod", klog.KObj(pod), "node", nodeName)
+		err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			// get latest version of the reservation
+			r, err1 := p.lister.Reservations(pod.Namespace).Get(pod.Name)
+			if errors.IsNotFound(err1) {
+				klog.V(3).InfoS("reservation not found, abort the update", "reservation", GetReservationNamespacedName(r))
+				return fmt.Errorf("reservation not found")
+			} else if err1 != nil {
+				klog.Warningf("failed to get reservation %v, err: %v", GetReservationNamespacedName(r), err1)
+				return err1
+			}
+
+			// check if the reservation has been expired
+			if r.Status.Phase == schedulingv1alpha1.ReservationExpired || p.reservationCache.IsExpired(GetReservationKey(r)) {
+				return fmt.Errorf("reservation is expired")
+			}
+
+			// mark reservation as available
+			setReservationAvailable(r, nodeName)
+			_, err1 = p.client.Reservations(r.Namespace).UpdateStatus(context.TODO(), r, metav1.UpdateOptions{})
+			// TBD: we may just set status.nodeName to avoid multiple updates
+			if err1 == nil {
+				_, err1 = p.client.Reservations(r.Namespace).Update(context.TODO(), r, metav1.UpdateOptions{})
+			} else {
+				klog.Warningf("failed to update status for reservation %v, err: %v", GetReservationNamespacedName(r), err1)
+			}
+			if err1 == nil {
+				return nil
+			}
+			klog.Warningf("failed to update reservation %v, err: %v", GetReservationNamespacedName(r), err1)
+			return err1
+		})
+		if err != nil {
+			return framework.NewStatus(framework.Error, "failed to bind reservation, err: "+err.Error())
+		}
+		return nil
+	}
+
+	return framework.NewStatus(framework.Skip, SkipReasonNotReservation)
+}
+
+func (p *Plugin) gcReservations() {
```

Comment:
It is recommended to separate these GC logic into a separate go file.

Reply:
done